pyfarm/agent/http/core/template.py

```diff
@@ -55,32 +55,25 @@ def dump_bytecode(self, bucket):  # pragma: no cover
         self.cache[bucket.key] = cache
 
 
-class DeferredTemplate(Template):
+class EncodedStringTemplate(_Template):
     """
-    Overrides the default :class:`.PackageLoader` so we
-    can produced the rendered result as a deferred call.
+    A sublcass of :class:`jinja2.Template which always returns
+    a string in response to :meth:`render`.  When Twisted writes
+    a response it always expects a bytes object.  In Python 2.x this
+    means we have to encode as a string.
     """
     def render(self, *args, **kwargs):
-        deferred = Deferred()
-
-        try:
-            # get the results then convert to a string, Twisted can't handle
-            # unicode from here
-            deferred.callback(
-                str(super(DeferredTemplate, self).render(*args, **kwargs)))
-        except Exception as e:  # pragma: no cover
-            deferred.errback(e)
-
-        return deferred
+        result = super(EncodedStringTemplate, self).render(*args, **kwargs)
+        return result.encode("ascii", "replace")
```

Comment:
Rolled back in e13c5fc757ee5a4d3533c804d4033f99f421519a.  This was in response to a behavior that the `DummyRequest` class in `write()`.